superset-frontend/src/explore/reducers/getInitialState.js

```diff
@@ -21,27 +21,40 @@ import shortid from 'shortid';
 import getToastsFromPyFlashMessages from '../../messageToasts/utils/getToastsFromPyFlashMessages';
 import { getChartKey } from '../exploreUtils';
 import { getControlsState } from '../store';
-import { getFormDataFromControls } from '../controlUtils';
+import {
+  getFormDataFromControls,
+  applyMapStateToPropsToControl,
+} from '../controlUtils';
 
 export default function getInitialState(bootstrapData) {
-  const controls = getControlsState(bootstrapData, bootstrapData.form_data);
-  const rawFormData = { ...bootstrapData.form_data };
-
+  const { form_data: rawFormData } = bootstrapData;
+  const slice = bootstrapData.slice;
+  const sliceName = slice ? slice.slice_name : null;
   const bootstrappedState = {
     ...bootstrapData,
+    sliceName,
     common: {
       flash_messages: bootstrapData.common.flash_messages,
       conf: bootstrapData.common.conf,
     },
     rawFormData,
-    controls,
     filterColumnOpts: [],
     isDatasourceMetaLoading: false,
     isStarred: false,
+    isInitializing: true,
   };
+  bootstrappedState.controls = getControlsState(bootstrappedState, rawFormData);
 
-  const slice = bootstrappedState.slice;
-  const sliceName = slice ? slice.slice_name : null;
+  // apply initial mapStateToProps for all controls, must execute AFTER
+  // bootstrappedState has initialized `controls`. Order of execution is not
```

Comment:
why order of execution is not guaranteed?

Reply:
I thought `Object.keys(...)` does not guarantee property order? Even if it does it's probably not a good idea to rely on it anyway.